### PR TITLE
EOS-20329: Cli framework movement

### PR DESCRIPTION
**Pull request by udayan-y21**
# Problem Statement
- Move cli framework to foundation repository, to consumed by other components

# Design  
- https://seagate-systems.atlassian.net/wiki/spaces/PRIVATECOR/pages/312705395/CLI+framework+to+Foundation+repository

# Coding 
-  Coding conventions are followed and code is consistent [Y/N]: Y
-  Confirm All CODACY errors are resolved [Y/N]: Y

# Testing 
Tested on: 

http://eos-jenkins.colo.seagate.com/job/Cortx-Deployment/job/VM-Deployment-R2-1Node/850/
http://cortx-storage.colo.seagate.com/releases/cortx/github/integration-custom-ci/centos-7.8.2003/custom-build-1471

- [x] Confirm that Test Cases are added (for both the cases, fix and feature) [Y/N]: Y
- [x] Confirm Test Cases cover Happy Path, Non-Happy Path and Scalability [Y/N]: Y
- [x] Confirm Testing was performed with installed RPM [Y/N]:  Y

# Review Checklist 
  Before posting the PR please ensure
- [x] PR is self reviewed
- [x] Jira is updated
- [x] Check if the description is clear and explained. 
- [x] Check Acceptance Criterion is defined. 
- [x] All the tests performed should be mentioned before Resolving a JIRA. 
- [x] Verification needs to be done before marked as Closed/Verified 

# Documentation
- [x] Changes done to WIKI / Confluence page

### CODE REVIEW

File: py-utils/src/setup/utils.py [hunk 1]
```diff
@@ -40,11 +41,18 @@ def __str__(self):
 class Utils:
     """ Represents Utils and Performs setup related actions """
 
+    @staticmethod
+    def _configure_utils_install_path():
+        with open(r'/etc/cortx/cluster.conf', 'w+') as file:
+            json.dump({}, file, indent=2)
+        Conf.load("conf_index", "json:///etc/cortx/cluster.conf")
+        Conf.set("conf_index", "install_path", "/opt/seagate/")
+        Conf.save("conf_index")
```
Comment (rahul-tripathi-git) [inline py-utils/src/setup/utils.py:50]: A sample file for /etc/cortx/cortx.conf with install_path in it is checked-in along with utils code as a part of another work item.
This method might be removed.

File: py-utils/src/setup/utils.py [hunk 2]
```diff
@@ -141,6 +149,9 @@ def init():
     def config(conf_url):
         """ Performs configurations """
 
+        # Configure cortx utils setup path.
+        Utils._configure_utils_install_path()
+
```
Comment (rahul-tripathi-git) [inline py-utils/src/setup/utils.py:154]: Same for this call.

Comment (rahul-tripathi-git) [reply py-utils/src/setup/utils.py:154]: Please directly use it like: (pseudo-code)
Conf.load("conf_index", "json:///etc/cortx/cluster.conf")
Conf.get("conf_index", "install_path")

**Review by rahul-tripathi-git — CHANGES_REQUESTED** [review]

File: py-utils/src/setup/utils.py (hunk 1 above)
Comment (udayan-y21) [reply py-utils/src/setup/utils.py]: Okay removed. Please review.

File: py-utils/src/setup/utils.py (hunk 2 above)
Comment (udayan-y21) [reply py-utils/src/setup/utils.py]: Removed this part of code. Please review

**Review by soniyamoholkar — APPROVED** [review]

**Review by pawankumarsrivastava — APPROVED** [review]

**Review by kupranay — APPROVED** [review]

**Review by kupranay — APPROVED** [review]

**Review by sachinpunadikar — APPROVED** [review]
Looks okay to me

**Comment by sachinpunadikar** [comment]
test this please

File: py-utils/src/utils/cli_framework/command.py [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+        :return: None
+        """
+        if add_parser_flag:
+             sub_parser = sub_parser.add_parser(data["name"],
+                        help=data["description"])
+        parser = sub_parser.add_subparsers()
+        for each_data in data["sub_commands"]:
+            self.add_args(each_data, parser, name)
+
+    def check_permissions(self, sub_command):
```
Comment (ujjwalpl) [inline py-utils/src/utils/cli_framework/command.py:120]: All the internal interfaces must start with _

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/validator/cli_validators.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+# For any questions about this software or licensing,
+# please email opensource@seagate.com or cortx-questions@seagate.com.
+
+import errno
+from cortx.utils.schema import payload
+from cortx.utils.log import Log
+from cortx.utils.cli_framework.errors import CliError
+
+class CliValidators:
```
Comment (ujjwalpl) [inline py-utils/src/utils/validator/cli_validators.py:21]: 1. Please follow naming conventions of the validator classes. Please refer to the other classes. 
2. Also name should singular.
3. There is nothing about CLI validation in this class. It should be called "FormatV" as its validating format. 
    Interface name should be "validate(format_type)"
4. There is no public interface to this class. There needs to be at least one public interface.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/cli_framework/command.py [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+                                  self.output)
+            err.write(f"{errstr}\n" or "")
+            return None
+        elif output_type:
+            output = getattr(Output, f"dump_{output_type}")(self.output,
+                                                            **kwargs)
+            out.write(f"{output}\n")
+
+    @staticmethod
+    def dump_success(output: Any, **kwargs):
```
Comment (ujjwalpl) [inline py-utils/src/utils/cli_framework/command.py:214]: return str.

File: py-utils/src/utils/cli_framework/command_factory.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+import argparse
+import sys
+import os
+from cortx.utils.schema.payload import Json
+from cortx.utils.cli_framework import const
+from cortx.utils.cli_framework.command import CommandParser
+
+
+class ArgumentParser(argparse.ArgumentParser):
```
Comment (ujjwalpl) [inline py-utils/src/utils/cli_framework/command_factory.py:24]: This class should be moved with CommandParser class preferably in a separate file called parser.py

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/cli_framework/command.py (hunk 5 above)
Comment (udayan-y21) [reply py-utils/src/utils/cli_framework/command.py]: Modifications done

File: py-utils/src/utils/validator/cli_validators.py (hunk 4 above)
Comment (udayan-y21) [reply py-utils/src/utils/validator/cli_validators.py]: > 1. Please follow naming conventions of the validator classes. Please refer to the other classes.

Done

> 2. Also name should singular. 

Done. 

> 3. There is nothing about CLI validation in this class. It should be called "FormatV" as its validating format.
Interface name should be "validate(format_type)"

Class name changed to FormatV. 
"validate(format_type, args)' --> currently framework is not supporting multiple arguments.
SO kept as `_validate_positive_int(value)` and `_validate_file_format(value)`

> 4. There is no public interface to this class. There needs to be at least one public interface.

Didnt understand, what do you mean "public interface"

File: py-utils/src/utils/cli_framework/command.py (hunk 3 above)
Comment (udayan-y21) [reply py-utils/src/utils/cli_framework/command.py]: Modifications done

File: py-utils/src/utils/validator/cli_validators.py (hunk 4 above)
Comment (udayan-y21) [reply py-utils/src/utils/validator/cli_validators.py]: @ujjwalpl Added public interface validate() 
Please review

File: py-utils/src/utils/validator/v_format.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    """
+
+    def validate(self, v_type, value):
+        if v_type == "positive_int":
+            FormatV._validate_positive_int(value)
+        elif v_type == "file_format":
+            FormatV._validate_file_format(value)
+
+    @staticmethod
+    def _validate_positive_int(value):
```
Comment (ujjwalpl) [inline py-utils/src/utils/validator/v_format.py:35]: Remove this. This is a simple one line if condition. You dont need validations. Validations are needed for subsystems.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/cli_framework/terminal.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    @staticmethod
+    def logout_alert(is_logged_out: bool):
+        if is_logged_out:
+            sys.stdout.write("Successfully logged out\n")
+        else:
+            Log.error(traceback.format_exc())
+            sys.stderr("Logout failed\n")
+
+    @staticmethod
+    def get_current_password(value):
```
[the remark this answers is not in the document]
Comment (udayan-y21) [reply py-utils/src/utils/cli_framework/terminal.py:61]: This is called from cortxcli 
```
          "flag": "-cp",
          "dest": "current_password",
          "type": "Terminal.get_current_password",
          "type_target": "cortx.utils.cli_framework.terminal",
          "suppress_help": true,
          "help": "",
```

File: py-utils/test/cli_framework/cli_test_data/users.json [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    "sub_commands": [
+      {
+        "name": "show",
+        "description": "Shows user list",
+        "need_confirmation": false,
+        "permissions_tag": "list",
+        "args": [
+          {
+            "flag": "-o",
+            "type": "CliValidators._positive_int",
```
[the remark this answers is not in the document]
Comment (udayan-y21) [reply py-utils/test/cli_framework/cli_test_data/users.json]: Modifications done

File: py-utils/src/utils/validator/v_format.py (hunk 7 above)
Comment (udayan-y21) [reply py-utils/src/utils/validator/v_format.py]: Modifications done

File: py-utils/src/utils/cli_framework/arg_validate.py [hunk 10]
```diff
@@ -0,0 +1,19 @@
+from cortx.utils.cli_framework.errors import CliError
```
Comment (ujjwalpl) [inline py-utils/src/utils/cli_framework/arg_validate.py:1]: Change the file name to arg_type or arg_validator.py

File: py-utils/src/utils/cli_framework/arg_validate.py [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+import errno
+
+
+class ArgType:
+
```
Comment (ujjwalpl) [inline py-utils/src/utils/cli_framework/arg_validate.py:6]: Add description to the every class you declare.

File: py-utils/src/utils/cli_framework/arg_validate.py (hunk 10 above)
Comment (ujjwalpl) [inline py-utils/src/utils/cli_framework/arg_validate.py:1]: license header missing

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/cli_framework/parser.py [hunk 12]
```diff
@@ -0,0 +1,126 @@
+import argparse
+import sys
+from cortx.utils.cli_framework import const
+from copy import deepcopy
+from importlib import import_module
+from cortx.utils.cli_framework.command import Command
+from typing import Dict, Any
+
+
```
Comment (ujjwalpl) [inline py-utils/src/utils/cli_framework/parser.py:9]: There should be exactly two blank lines.

File: py-utils/src/utils/cli_framework/command_factory.py [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+        CommandFactory.edit_arguments(namespace)
+
+        sys_module = sys.modules[__name__]
+        for attr in ["command", "action", "args"]:
+            setattr(sys_module, attr, getattr(namespace, attr))
+            delattr(namespace, attr)
+        return command(action, vars(namespace), args)
+
+    @staticmethod
+    def edit_arguments(namespace):
```
Comment (ujjwalpl) [inline py-utils/src/utils/cli_framework/command_factory.py:68]: This is internal function and must start with _

File: py-utils/test/cli_framework/cli_test_data/users_show_table_output [hunk 14]
```diff
@@ -0,0 +1,7 @@
++-----------+-----------+---------------------+-----------------------------------+-----------------------------------+
```
Comment (ujjwalpl) [inline py-utils/test/cli_framework/cli_test_data/users_show_table_output:1]: Please rename cli_test_data to test_data. Its already qualified by parent.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/test/cli_framework/cli_test_data/users.json [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+    "sub_commands": [
+      {
+        "name": "show",
+        "description": "Shows user list",
+        "need_confirmation": false,
+        "permissions_tag": "list",
+        "args": [
+          {
+            "flag": "-o",
+            "type": "IntType.validate",
```
Comment (ujjwalpl) [inline py-utils/test/cli_framework/cli_test_data/users.json:13]: You should only specify type here, i.e. 
    "type": "IntType"

.validate should be suffixed during processing.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/test/cli_framework/cli_test_data/users_show_table_output (hunk 14 above)
Comment (udayan-y21) [reply py-utils/test/cli_framework/cli_test_data/users_show_table_output]: Modifications done

File: py-utils/src/utils/cli_framework/command_factory.py (hunk 13 above)
Comment (udayan-y21) [reply py-utils/src/utils/cli_framework/command_factory.py]: Modifications done

File: py-utils/src/utils/cli_framework/parser.py (hunk 12 above)
Comment (udayan-y21) [reply py-utils/src/utils/cli_framework/parser.py:24]: Modifications done

File: py-utils/src/utils/cli_framework/arg_validate.py (hunk 10 above)
Comment (udayan-y21) [reply py-utils/src/utils/cli_framework/arg_validate.py]: Modifications done

File: py-utils/src/utils/cli_framework/arg_validate.py (hunk 11 above)
Comment (udayan-y21) [reply py-utils/src/utils/cli_framework/arg_validate.py]: Modifications done

File: py-utils/src/utils/cli_framework/arg_validate.py (hunk 10 above)
Comment (udayan-y21) [reply py-utils/src/utils/cli_framework/arg_validate.py]: Modifications done

File: py-utils/test/cli_framework/cli_test_data/users.json (hunk 15 above)
Comment (udayan-y21) [reply py-utils/test/cli_framework/cli_test_data/users.json]: Evalution of `"type"` and `"type_target"` is done in common place in framework code.. as below
```python
                   if each_args.get("type_target"):
                        module_obj = import_module(each_args.pop("type_target"))
                        each_args["type"] = eval(f"module_obj.{each_args['type']}")
                    else:
                        each_args["type"] = eval(each_args["type"])
```
This part of code executes at various place.. some of the examples are: 
```
          "type": "FormatV.validate",
          "type_target": "cortx.utils.validator.v_format",
```
```
          "type": "IntType.validate",
          "type_target": "cortx.utils.cli_framework.arg_validator",
```
```          
          "type": "Terminal.get_password",
          "type_target": "cortx.utils.cli_framework.terminal"
```
So adding class name `(IntType)` to `"type"` will not work...

File: py-utils/src/utils/cli_framework/arg_validator.py [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+
+from cortx.utils.cli_framework.errors import CliError
+import errno
+
+
+class ArgType:
+    """
+    Base class for Argumentent type validaton
+    """
+    def validate(self, val):
```
Comment (ujjwalpl) [inline py-utils/src/utils/cli_framework/arg_validator.py:24]: @staticmethod

File: py-utils/test/cli_framework/test_data/users.json [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+    "sub_commands": [
+      {
+        "name": "show",
+        "description": "Shows user list",
+        "need_confirmation": false,
+        "permissions_tag": "list",
+        "args": [
+          {
+            "flag": "-o",
+            "type": "IntType.validate",
```
Comment (ujjwalpl) [inline py-utils/test/cli_framework/test_data/users.json:13]: "type": "IntType",

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/test/cli_framework/cli_test_data/users.json (hunk 15 above)
Comment (ujjwalpl) [reply py-utils/test/cli_framework/cli_test_data/users.json]: each_args["type"] = eval(f"module_obj.{each_args['type']}.validate")

File: py-utils/test/cli_framework/test_data/users.json (hunk 17 above)
Comment (udayan-y21) [reply py-utils/test/cli_framework/test_data/users.json:13]: Please check my reply https://github.com/Seagate/cortx-utils/pull/303#discussion_r645379133

File: py-utils/src/utils/cli_framework/arg_validator.py (hunk 16 above)
Comment (udayan-y21) [reply py-utils/src/utils/cli_framework/arg_validator.py:25]: Modifications done

File: py-utils/test/cli_framework/test_data/users.json (hunk 17 above)
Comment (ujjwalpl) [reply py-utils/test/cli_framework/test_data/users.json:13]: We need to be consistent in the CLI Spec. 
"type": "<Has to be type>"
"validator": "<Has to be validator>"

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/test/cli_framework/cli_test_data/users.json (hunk 15 above)
Comment (udayan-y21) [reply py-utils/test/cli_framework/cli_test_data/users.json]: @ujjwalpl Modifications done
As per discussion with Pranay and Pawan,
We have decided that Validations will be handled by api for which its called(seperate ticket will be created for that).
As these validations should  be the part of functionality underlying to it and not framework. 
 
CLI-framework will provide
**type**: <defined-data-type, ex: int, str, etc>,
**type_method** : < specific method of class to be executed for that arg>,
**type_target**:< import path of type_method>,


**Basic validations** like ( int is int, str is str, etc) is handled by Argparse library itself
```
Ex:
cortxcli$ users show -o apple
usage: cortxcli.py users show [-h] [-o OFFSET] [-l LIMIT]
                              [-s {user_id,user_type,created_time,updated_time}]
                              [-d {desc,asc}] [-f {table,xml,json}]
Error: Argument -o: invalid int value: 'apple'
cortxcli$
 
cortxcli$ users show -o 1.1
usage: cortxcli.py users show [-h] [-o OFFSET] [-l LIMIT]
                              [-s {user_id,user_type,created_time,updated_time}]
                              [-d {desc,asc}] [-f {table,xml,json}]
Error: Argument -o: invalid int value: '1.1'
```

File: py-utils/test/cli_framework/test_data/users.json (hunk 17 above)
Comment (udayan-y21) [reply py-utils/test/cli_framework/test_data/users.json]: Please review

File: py-utils/src/utils/cli_framework/cli_client.py [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+        self._url = url
+
+    def call(self, command):
+        pass
+
+    def process_request(self, session, cmd, action, options, args, method):
+        pass
+
+
+class DirectClient(Client):
```
Comment (ujjwalpl) [inline py-utils/src/utils/cli_framework/cli_client.py:33]: Rename this class to CliClient if this meant for CLI.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/cli_framework/terminal.py [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+        """
+        Fetches the Password from Terminal in Non-Echo Mode.
+        :return:
+        """
+        sys.stdout.write(("\nPassword must contain the following.\n1) 1 upper and lower "
+        "case character.\n2) 1 numeric character.\n3) 1 of the !@#$%^&*()_+-=[]{}|' "
+                          "characters.\n"))
+        value = value or getpass(prompt="Password: ")
+        if not value:
+            raise CliError(errno.EINVAL, Terminal.EMPTY_PASS_FIELD)
```
Comment (ujjwalpl) [inline py-utils/src/utils/cli_framework/terminal.py:84]: For consistency please replace  EMPTY_PASS_FIELD with actual message. This is the only place where it is used.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/cli_framework/cli_client.py [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+import aiohttp
+
+
+class Client:
+    """ Base class for invoking business logic functionality """
+
+    def __init__(self, url):
+        self._url = url
+
+    def call(self, command):
```
Comment (ujjwalpl) [inline py-utils/src/utils/cli_framework/cli_client.py:26]: Merge call and process_request. They are the same thing. Derived classes should implement the same interface.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/cli_framework/cli_client.py (hunk 20 above)
Comment (udayan-y21) [reply py-utils/src/utils/cli_framework/cli_client.py]: Modifications done

File: py-utils/src/utils/cli_framework/terminal.py (hunk 19 above)
Comment (udayan-y21) [reply py-utils/src/utils/cli_framework/terminal.py]: Modifications done

File: py-utils/src/utils/cli_framework/cli_client.py [hunk 21]
```diff
@@ -0,0 +1,58 @@
+# CORTX-Py-Utils: CORTX Python common library.
```
[the remark this answers is not in the document]
Comment (udayan-y21) [reply py-utils/src/utils/cli_framework/cli_client.py]: Modifications done

File: py-utils/src/utils/cli_framework/cli_client.py (hunk 18 above)
Comment (udayan-y21) [reply py-utils/src/utils/cli_framework/cli_client.py]: Modifications done

**Review by ujjwalpl — APPROVED** [review]

**Review by rahul-tripathi-git — APPROVED** [review]
Looks good to me.

**Pull request merged**